Project 9

fostert14 opened the pull request.

fostert14 committed:
Complete Deploy Project Steps

fostert14 committed:
Create Api.js/Finish step 3

fostert14 committed:
Complete Step 1

fostert14 committed:
Complete step 2

fostert14 committed:
Complete step 3

fostert14 committed:
Complete step 4

fostert14 committed:
Create confirm delete modal

fostert14 committed:
Complete Step 5

fostert14 committed:
Complete Step 6

fostert14 committed:
Complete Step 7

fostert14 committed:
Create Edit  Profile Picture Hover Button

fostert14 committed:
Create Edit Profile Image Modal

fostert14 committed:
Add Ability to Edit Avatar/Complete Step 8

fostert14 committed:
Fix card rendering order, new card event listeners

fostert14 committed:
Complete Step 9

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,82 @@
+  getInitialCards() {
+    return fetch(`${this._url}/cards`, {
+      headers: this._headers,
+    }).then(this._checkResponse);
```
**Could be improved**

You can make a special method for fetching and checking responses **not to duplicate** it in every request:

```js

  _request(url, options) {
    return fetch(url, options).then(this._checkResponse)
  }

```

And now you can safely replace `fetch` with `this._request` and remove repeated checking. And you don’t need to change anything else.

gennady-bars commented on src/components/PopupWithDelete.js:
```diff
@@ -0,0 +1,32 @@
+  }
+
+  open(cardElement, cardID) {
+    this._cardElement = cardElement;
+    this._cardID = cardID;
```
nice way of updating the data

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -28,6 +29,14 @@ export default class PopupWithForm extends Popup {
 
+  renderLoading(isLoading) {
+    if (isLoading) {
+      this._button.textContent = "Saving...";
```
**Could be improved**

It’s better to make method `renderLoading` inside class `PopupWithForm`. You should find the button once in the constructor and change the text:  

```js
  constructor({ popupSelector, handleSubmit }) {
    super(popupSelector);
    // fix the initial button text only once in the constructor
    this._submitBtnText = this._submitBtn.textContent
  }
  // add 2 params: isLoading and loadingText with a default text
  renderLoading(isLoading, loadingText='Saving...') {
    if (isLoading) {
      this._submitBtn.textContent = loadingText;
    } else {
	// here we return back the initial text. So, you don’t need to bother yourself about it
      this._submitBtn.textContent = this._submitBtnText;
    }
  }

```

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -28,6 +29,14 @@ export default class PopupWithForm extends Popup {
+      this._button.textContent = "Save";
+    }
+  }
+
```
**Could be improved**

You can make ` setInputValues` in class `PopupWithForm`, which can insert data into inputs:

```js

  setInputValues(data) {
    this._inputList.forEach((input) => {
      // here you insert the `value` by the `name` of the input
      input.value = data[input.name];
    });
  }



```

So, this way you’ll not have to search the inputs of the profile in `index.js`

gennady-bars commented on src/pages/index.js:
```diff
@@ -30,17 +34,69 @@ const enableValidation = (settings) => {
+    .addNewCard({ name, link })
+    .then((newCardData) => {
+      const { name, link } = newCardData;
+      cardSection.addItem(
+        new Card(
+          newCardData,
+          "#card-template",
+          imagePopup,
+          (title, link) => imagePopup.open(title, link),
+          (cardElement, cardID) => {
+            deleteImagePopup.open(cardElement, newCardData._id);
+          },
+          api
+        ).getView()
```
You have `renderCard` for creating cards. Please, don't duplicate the code

gennady-bars commented on src/pages/index.js:
```diff
@@ -30,17 +34,69 @@ const enableValidation = (settings) => {
+      );
+    })
+    .catch((err) => console.error("Error adding new card:", err));
   addImagePopup.close();
```
`  addImagePopup.close();
`

You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well


**Please, correct it everywhere**

gennady-bars commented on src/pages/index.js:
```diff
@@ -30,17 +34,69 @@ const enableValidation = (settings) => {
-    name: name,
-    job: description,
-  });
+  editProfilePopup.renderLoading(true);
```
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```

fostert14 replied:
Implemented. Took me a second to understand, but I like it. Thanks for the suggestion!

gennady-bars commented on src/pages/index.js:
```diff
@@ -30,17 +34,69 @@ const enableValidation = (settings) => {
+  api
+    .updateAvatar(link)
+    .then((newAvatar) => {
+      avatarIMG.src = newAvatar.avatar;
```
`      avatarIMG.src = newAvatar.avatar;
`
* Class `UserInfo` should set the entire user’s data (including the **avatar**)

gennady-bars commented on src/pages/index.js:
```diff
@@ -54,10 +110,58 @@ const renderCard = (cardData) => {
+      name: userData.name,
+      job: userData.about,
+    });
+    avatarIMG.src = userData.avatar;
```
* Class `UserInfo` should set the entire user’s data (including the **avatar**)

gennady-bars commented on src/pages/index.js:
```diff
@@ -54,10 +110,58 @@ const renderCard = (cardData) => {
+    cardSection.renderItems();
+  })
+  .catch((err) => {
+    console.error("Error fetching cards:", err);
```
Good job that you `catch` possible errors at the end of server requests

gennady-bars commented on src/pages/index.js:
```diff
@@ -30,17 +34,69 @@ const enableValidation = (settings) => {
+    })
+    .catch((err) => {
+      console.error("Error deleting card:", err);
+    });
```
* When I submit the form for the 1st time the submit button gets `Saving...` for always. You need to return the **default** button text only in blocks `finally` so that the button could get the text back in any case.

gennady-bars reviewed (commented):
Dear Trevor , (**Please, expand the general comment ↓**)

Good job but there are some remarks:



* When I submit the form for the 1st time the submit button gets `Saving...` for always. You need to return the **default** button text only in blocks `finally` so that the button could get the text back in any case. 
* When I reload the page the black hearts become white again
* Class `UserInfo` should set the entire user’s data (including the **avatar**)
* You have `renderCard` for creating cards. Please, don't duplicate the code
* You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well


Please, correct everything and your project will be accepted.

Good luck with the refactoring

fostert14 committed:
Add: UserInfo sets Avatar

fostert14 committed:
Refactor api.js request into _request method

fostert14 committed:
Refactor Code

PopupWithDelete more streamlined,
setInputValues in PopupWithForm added.

fostert14 committed:
Refactor handleImageFormSubmit

now uses renderCard function.

fostert14 committed:
Refactor handleImageForm Submit

Added new handleSubmit function

refactored so all "description/job/about" sections are now just "about"

fostert14 committed:
Refactor all form submissions

All refactor submissions now use the new handleSubmit function.

fostert14 committed:
Fix: Like Button

Like button now stays the appropriate shade when reloading the page.

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -29,14 +31,20 @@ export default class PopupWithForm extends Popup {
     return inputValues;
```
Please, use `this._inputList ` in line 27 not to seach them again in line 25


```js
  _getInputValues() {
    const inputValues = {};
    for (const input of this._inputList) {  // here
      inputValues[input.name] = input.value;
    }

    return inputValues;
  }

```

gennady-bars commented on src/pages/index.js:
```diff
@@ -34,75 +44,49 @@ const enableValidation = (settings) => {
         name: updatedUser.name,
-        job: updatedUser.about,
+        about: updatedUser.about,
+        avatar: userInfo.getUserInfo().avatar,
```
`        avatar: userInfo.getUserInfo().avatar,
`

You have everything in `updatedUser` (including the avatar)


```js

      userInfo.setUserInfo(updatedUser);

```

fostert14 replied:
I've fixed this part. I did run into a similar issue in my handleAvatarSubmit, but don't call updatedUser there since it returns the avatar link directly. Should I use userInfo.getUserInfo() in this case?

gennady-bars replied:
> Should I use userInfo.getUserInfo() in this case?

`userInfo.getUserInfo()` is not needed there because you get the whole object from the server with all 4 params (including avatar)

gennady-bars commented on src/pages/index.js:
```diff
@@ -34,75 +44,49 @@ const enableValidation = (settings) => {
+  function makeRequest() {
+    return api.updateAvatar(link).then((newAvatar) => {
+      userInfo.setUserInfo({
+        avatar: newAvatar.avatar,
```
* When I change the avatar the whole profile data disappears https://disk.yandex.ru/i/4ehw26tajPB5vw

gennady-bars reviewed (commented):
One more step and everything will be fine:

* When I change the avatar the whole profile data disappears https://disk.yandex.ru/i/4ehw26tajPB5vw

fostert14 committed:
Refactor: Card.js:

Remove this._likeButton from setEventListeners method in Card.js, as it is now in getView

fostert14 committed:
Refactor _getInputValues

fostert14 committed:
Refactor handleProfilesbmt - avatar

gennady-bars commented on src/pages/index.js:
```diff
@@ -67,6 +67,8 @@ function handleAvatarSubmit(link) {
   function makeRequest() {
     return api.updateAvatar(link).then((newAvatar) => {
       userInfo.setUserInfo({
+        name: userInfo.getUserInfo().name,
+        about: userInfo.getUserInfo().about,
```
`userInfo.getUserInfo()` is not needed here because you get the whole object from the server with all 4 params (including avatar)


```js
    return api.updateAvatar(link).then((userData) => {
      userInfo.setUserInfo(userData);
    });

```


Please, call it `userData` instead of `newAvatar` because the response has the whole data of the user

fostert14 replied:
Ah cheers! Thank you so much for your help here and the rest of my project 👍

gennady-bars reviewed (approved):
You’ve done a great job.

Your project has been accepted.

Good luck

fostert14 committed:
Refactor HandleAvatarSbmt Function
